src/Streamly/Internal/Data/Parse/Types.hs

```diff
@@ -0,0 +1,170 @@
+-- drive a parser we call the @initial@ action to retrieve the initial state of
+-- the fold. The driver invokes @step@ with the state returned by the previous
+-- step and the next input element. It results into a new state and a command
+-- to the driver represented by 'Step' type. At any point of time the driver
+-- can call @extract@ to inspect the result of the fold. It may result in an
+-- error or an output value.
+--
+data Parser m a b =
+    forall s. Parser (s -> a -> m (Step s)) (m s) (s -> m (Either String b))
+
+-- | Maps a function over the output of the fold.
+--
+instance Monad m => Functor (Parser m a) where
+    {-# INLINE fmap #-}
+    fmap f (Parser step initial extract) =
+        Parser step initial (fmap3 f extract)
+
+        where
+
+        fmap3 g = fmap (fmap (fmap g))
+
+    {-# INLINE (<$) #-}
+    (<$) b = \_ -> pure b
+
+{-# ANN type SeqParseState Fuse #-}
+data SeqParseState sl f sr =
+    SeqParseL sl | SeqParseR f sr | SeqParseLErr String
+
+-- | Apply two parsing folds sequentially to an input. The input is provided to
+-- the first parser, if the parser fails then whole composition fails. If the
+-- parser succeeds, the remaining input is supplied to the second parser. If
+-- the second parser succeeds the composed parser succeeds otherwise it fails.
+--
+instance Monad m => Applicative (Parser m a) where
+    {-# INLINE pure #-}
+    pure b = Parser (\_ _ -> return $ Stop 0 ())
+                    (return ())
+                    (\_ -> return $ Right b)
+
+    {-# INLINE (<*>) #-}
+    (Parser stepL initialL extractL) <*> (Parser stepR initialR extractR) =
+        Parser step initial extract
+
+        where
+
+        initial = SeqParseL <$> initialL
+
+        -- Note: For the composed parse to terminate, the left parser has to be
+        -- a terminating parser returning a Halt at some point.
+        step (SeqParseL st) a = do
+            r <- stepL st a
+            case r of
+                Yield n s -> return $ Yield n (SeqParseL s)
+                Skip n s -> return $ Skip n (SeqParseL s)
+                Stop n s -> do
+                    res <- extractL s
+                    case res of
+                        Left err -> return $ Stop n (SeqParseLErr err)
+                        Right f -> Skip n <$> (SeqParseR f <$> initialR)
+
+        step (SeqParseR f st) a = do
+            r <- stepR st a
+            return $ case r of
+                Yield n s -> Yield n (SeqParseR f s)
+                Skip n s -> Skip n (SeqParseR f s)
+                Stop n s -> Stop n (SeqParseR f s)
+
+        step (SeqParseLErr _) _ = error "step called in SeqParseLErr"
+
+        -- XXX bimap to add "<*>: Right parse failed" to the error message
+        extract (SeqParseR f s) = fmap (fmap f) (extractR s)
+        extract (SeqParseL _) = return $ Left "<*>: Incomplete left parse"
+        extract (SeqParseLErr err) =
+            return $ Left $ "<*>: Left parse failed\n" ++ err
```

Comment:
Is `Parser m a` not a Monad?

Reply:
Not sure how a `Monad` can be written/useful, a `Comonad` is more likely to be useful.